### PR TITLE
[wip] Partial buys

### PR DESCRIPTION
**Pull request by es-fem**
Thank you for sending your pull request. But first, have you included
unit tests, and is your code PEP8 conformant? [More details](https://github.com/freqtrade/freqtrade/blob/develop/CONTRIBUTING.md)

## Summary
Explain in one sentence the goal of this PR

Solve the issue: #___

## Quick changelog

- <change log #1>
- <change log #2>

## What's new?
*Explain in details what this PR solve or improve. You can include visuals.*

### CODE REVIEW

**Comment by es-fem** [comment]
To try it add this functions to the strategy 

from freqtrade.strategy.interface import PartialTradeTuple as partial

def increase_position ( dataframe: pd.DataFrame, amount: float) -> pd.Series:
    return pd.Series([partial(1,amount) for x in range(len(dataframe))])

def reduce_position ( dataframe: pd.DataFrame, amount: float) -> pd.Series:
    return pd.Series([partial(1,amount) for x in range(len(dataframe))])

And call them like: (you will still need 'buy' and 'sell' set to 0)
dataframe.loc[
            (
                    (dataframe['k'] > dataframe['d']) &
                    (dataframe['k'] < dataframe['oversold']) &
                    (dataframe['d'] < dataframe['oversold'])

            ),
            'partial_buy'] = increase_position(dataframe, 15)

**Comment by xmatthias** [comment]
After having a very quick glance over this - i don't think it'll work well - at least not the way it is now.

----

To be honest, i really don't like the interface you're proposing, and i don't think it'll work well. 
Using an (optional) tuple instead of a boolean field is quite risky and errorprone, especially when non-programmers use this.
This interface method will most likely also not work correctly for backtesting.


The buy / sell methods seem to do lot's of code duplication (seem like copies of their existing counterparts, with slight modifications), and are way too long - which means they do way too much.

---

It'll also be highly problematic in longer timeframes (say 1h) ... 
the assumption is that a signal is be active for the whole candle (this is usually the case, in some strategies a signal can be active for multiple candles - if they're not using "crossed").

Unless i missed it, you're not protecting against this - so you'd buy another slice on every iteration until you run out of money, or the signal disappears (which is when the candle is over).

That's obviously very risky, and in my opinion, not useful at all.

**Comment by es-fem** [comment]
Well, its a first aproach.. I'll do this if you're agree:

- Change the tuple for two booleans
- unify buy/sell methods to save code
- lock the candle to avoid rebuys/resells
- Modify backtesting and hyperopt when the rest is done.

Im not good with db and for dry_run I had to tweak wallets to calculate
stake_amount. May be we should add a table in the db for the wallets?

**Comment by xmatthias** [comment]
I'm not sure i understand why you would need wallets in the database? 

all data necessary is in the trades (it'll also be that way in the future - otherwise you're tracking the wrong thing) - so a wallet table will mean one thing - the data will be there twice, which will eventually lead to inconsistencies.

---

As said, i'm not sure about the way you're doing the interface - it seems overly complicated and unnecessary.
There's a very simple way to allow position stacking:
add a new config entry which enables / disables position stacking.

* If it's disabled - nothing changes (this is how freqtrade works now).
* If it's enabled, you can buy "stake_amount" multiple times, which will then be combined into one trade (I'm not sure about combining positions - see https://github.com/freqtrade/freqtrade/issues/1519#issuecomment-659154906 for details)

That way, the strategy interface doesn't change - the whole code will be a lot cleaner.

That way you can't dynamically modify the amount you buy - but that's not necessary to enable position stacking.
Also, i would not do that in the dataframe, but as part of a callback function (which must be a seperate PR, otherwise the change would be too large to review).


Before doing any more work on the interface, you should maybe also read #2183. Either using a tuple, or anything else will limit the possibilities of introducing shorting or leveraged pairs - so it's definitely not a good idea.

**Comment by es-fem** [comment]
Hi, The wallets in the database idea is because if you modify the open
trade lets say selling some, the actual calcs don't actualize the wallet if
the trade is open. I had to modify the wallets.py to consider stake_amount
as part of the USD wallet for open trades, breaking the logic of
"stake_amount = amount * open_rate" in the db. (this  logic is also a
duplication thus open_rate and amount are already in the db so
stake_amount  could be calculated anytime).
I like the idea of #2183, I don't know if it will work with stake =
unlimited, but it's better than my idea.
The only thing that maybe is usable from my code is the method that
averages the open_price. That's a standard formula needed to calculate
unrealized p&l for a position plus leverage and shorts options are easy to
add  in p&l.  something like this:
https://help.bybit.com/hc/en-us/articles/900000404726-P-L-calculations-Inverse-
What I was trying to do is to emulate exchanges. They consider an open
position , modify contracts in the position and recalculate  average open
price when you add contracts to it without closing the trade. But combining
trades (as they are right now) to build a new actualized trade entry will
work as well  ( It looks more complicated to me but I understand it is
better for backward compatibility).

File: freqtrade/freqtradebot.py [hunk 1]
```diff
@@ -348,26 +347,28 @@ def enter_positions(self) -> int:
...
-            logger.info("No currency pair in active pair whitelist, "
-                        "but checking to sell open trades.")
-            return trades_created
-        # Create entity and execute trade for each pair from whitelist
-        for pair in whitelist:
-            try:
-                trades_created += self.create_trade(pair)
-            except DependencyException as exception:
-                logger.warning('Unable to create trade for %s: %s', pair, exception)
+        else:
```
Comment (mrsegen) [inline freqtrade/freqtradebot.py:350]: `else` unnecessary after an `if` that contains a return. ("Flat is better than nested" - [Zen](https://www.python.org/dev/peps/pep-0020/))

**Pull request closed without merging**